README.md
Fork from:

    Steam Launcher - Start Steam Big Picture Mode from within Kodi
    See Kodi thread for more details:
    http://forum.kodi.tv/showthread.php?tid=157499

    https://github.com/teeedubb
    http://store.steampowered.com/bigpicture
    http://kodi.tv/


changelog.txt
0.0.1

First version,

0.1.0

Reworked the windows side, now xbmc restarts once steam is exited.
Thanks to malte for Rom Collection Browser. http://forum.xbmc.org/showthread.php?tid=70115
Thanks to JohnnyD2 for VB scripts http://forum.xbmc.org/showthread.php?tid=70115&pid=960394#pid960394
I used their work to improve the addon.

0.1.1

Reworked linux side, now xbmc restarts after steam is closed.
Thanks to malte for Rom Collection Browser. http://forum.xbmc.org/showthread.php?tid=70115
Thanks to Resandro for Arcade Browser. http://forum.xbmc.org/showthread.php?tid=48983
I used their work to improve the addon

0.1.2

Minor change to allow install on osx. Other small changes.

0.1.3

Additions to Windows scripting: Xbmc will stay closed if Steam updates itself, it will wait for up to 5 seconds for Steam to restart. If steam is already open Xbmc will exit and Steam will switch to big picture mode.

0.1.4
Moved scripts into new directory.

0.1.5
Windows command windows now hidden.

0.1.6
Reworked Linux scripting - If steam is already open Xbmc will exit and Steam will switch to big picture mode.
Reorganised files.

0.1.7
Small fix

0.1.8
Linux now uses a lockfile to prevent multiple instances of xbmc reopening if Steam has been exited to desktop and xbmc re-opened.

0.1.9 
New windows scripting

0.2.0 
Windows checks whether its 32 or 64 bit.
New icon.

0.2.2
Changes to steam-launch.bat and default.py

0.3.4
AutoHotKey .exe used instead of batch file, now when exiting to desktop xbmc relaunches in windows. ahk script is included in case you want to make changes to the default timeouts, eg: start steam, wait X seconds then kill xbmc. The ahk script will force focus on either xbmc/steam when they are run. Script usage is SteamLaunch-AHK.exe "d:\games\steam\steam.exe" "d:\xbmc\xbmc.exe"
Fanart added.

0.3.6
Changes to AHK script - disable #persistence and force #singleinstance

0.4.0
Added back code to autodetect if windows is 32 or 64 bit.
New steam-launch.sh and now xbmc will restart if steam is exited or returned to desktop mode. The linux version now requires the program 'wmctrl' to work.

0.4.4
Small fixes.

0.5.0
Changes to windows side of default.py - now cmd window completely hidden and will not steal focus away from xbmc when in a true full screen.
Somewhere along the line the <provides> tag in addon.xml was changed to 'game'. If a xbmc build with a Games main menu is used (eg garbear's retroplayer builds), the addon will appear under games, if not it will appear under Programs.

0.5.1
Recompiled AHK script using 32bit AHK, fixes script not working on win32
Linux version now makes the steam-launch.sh script executable when run, no need to manually do it (it annoyed the hell out of me!). 

0.8.4
Test version.
Steam/XBMC paths now changed via addon settings gui.
Scripts are now launched from the addon_data directory, if the script does not exist it will be copied over from the addon install directory. Custom scripts will now not be deleted on addon update.
Removed unnecessary scripts plus changes to remaining scripts.
Plus other changes.
Big thanks to malte, I used code and ideas from Rom Collection Browser. Cheers!

0.8.6
Added 2 advanced settings:
1. Delete scripts and update on next run will delete the scripts out of the addon_data dir and copy them from the addon install dir the next time the addon is run. This is handy if you want to go back to the default scripts or if thebundled scripts have been updated. Only the script file will be deleted and the setting is turned off (to default) when run once.
2. Linux only: Make steam-launch.sh executable on next run - This changed the executable bit on the file 'steam-launch.sh' which is necessary for this addon to run. This setting defaults to off after being run once.
Addon now only copies/deletes os specific scripts.
Small changes

0.9.0
Re-added and made changes to launchhidden.vbs, it's needed to prevent 'unable to launch application' error messages from xbmc restarting too fast.
Changes to default.py

0.9.2
Changes to default.py
Check if script files exist before deleting, prevents a error.
Updated Steam-Lanuch-AHK.exe - Additional checks incase BPM is already active + for when BPM is restarted for forced/beta updates, script will check for 5 seconds for BPM or the steam update window to show up again, if it does it kills xbmc and waits for BPM to be closed. Also removed all hard timeouts and now everything happens on window events, eg XBMC is killed when either the BPM or steam update window appear.

0.9.3
Fixed wrong AHK.exe.
Added busy dialog when script is run.
Updated docs.

0.9.5
Cleaned up addon + updated script files for linux + windows. 
Readded macosx code.
Option to minimize xbmc instead of exiting it, this is set via addon settings. Yes, no and ask are available.
Checks if files steam/xbmc locations exist and/or steam-launch.sh is executable, If not a yes/no dialog appears which can take you to the settings.
Configurable timeout for 'Busy Dialog".

0.9.7
Checks for script updates and notifies user.

0.9.8
OSX now working, although xbmc will only restart after quitting steam completely.
Option to disable program file checks
Small changes.

0.9.9
Reverted windows changes in 0.9.8

0.9.10
French translation thanks to Kre|S.
OSX checks for XBMC + Steam paths.
A advanced option to use a Batch file to launch scripts - uses a batch file as in previous version which should hopefully solve the 'nothing happening' issue some people have been having. GUI settings for XBMC + Steam paths and whether to quit XBMC have no effect with this setting enabled, the options are configured via steam-launch.bat, located in userdata/addon_data/script.steam.launcher/scripts/ , with instructions located in file itself. steam-launch.bat is not deleted by the addon or updates and if it does not exist it will be copied to the userdata directory on first run with the option enabled, along with launchhidden-bat.vbs

resources/scripts/steam-launcher.sh
#!/bin/bash
#Script to launch Steam BPM from Kodi, by teeedubb
#See: https://github.com/teeedubb/teeedubb-xbmc-repo http://forum.kodi.tv/showthread.php?tid=157499
# Manual script usage:
#   Example : SteamLauncher-AHK.exe "e:\path\to\steam.exe" "d:\path\to\kodi.exe" "0/1" "true/false" "scriptpath/false" "scriptpath/false" "steam parameters" "true/false" 
#   Parameters:
#     $3 = 0 Quit KODI, 1 Minimize KODI.
#     $4 = KODI portable mode.
#     $5 = pre script.
#     $6 = post script.
#     $7 = steam parameters.
#     $8 = Force kill kodi after X seconds, 0 to disable.
#     $9 = start in desktop mode.
#Change the 'steam.launcher.script.revision =' number to 999 to preserve changes through addon updates, otherwise it shall be overwritten.
#steam.launcher.script.revision=017



# delete $9 $8
# ex: 
#   $1 = /usr/bin/steamlink
#   $2 = /usr/local/lib/kodi/kodi.bin
#   $3 = 0
#   $4 = true
#   $5 = false
#   $6 = false
#   $7 = 5
#   $8 = true

ME="$(pwd)/$0"
ME="$(realpath $ME)"

log(){
  echo "$*"
  echo "$*" >> $ME.log
}

startKodi(){
    kodi start &

    pkill -P $(pidof -x steamlink.sh)
    pkill -P $(pidof -x streaming_client)
}

if [[ "$1" -eq "forced" ]]; then
  shift
  log "let's go. $ME $*"
else
  echo "rerun independently" > $ME.log
  nohup $ME "forced" $* &
  kill -9 $$
fi


l=script.steam.launcher: 
if [ -z "$*" ]; then
  log $l "No arguments provided, see script file for details."
  exit
fi

case "$(uname -s)" in
    Linux)
      #
      KODI_BIN=$(ps aux | grep -E 'kodi.bin|kodi-x11|kodi-wayland|kodi-gbm' | grep -v 'grep.*kodi' | head -n1 | awk '{print $11}')
      log "$l find kodi process: $KODI_BIN"

      #pre script
      if [[ $5 != false ]] ; then
        "$5"
      fi

      #close kodi
      log "$l killing kodi"
      kodi stop
      (log $l forcefully killing kodi after "$8"s ; sleep $8 ; if [[ $(pidof $KODI_BIN) ]] ; then killall -9 $KODI_BIN ; fi)&

      #killing old
      log "$l killing old streamlink"
      if [[ $(pidof streaming_client) ]] ; then
         pkill -P $(pidof -x streaming_client)
      fi
      if [[ $(pidof steamlink.sh) ]] ; then
         pkill -P $(pidof -x steamlink.sh)
      fi

      log "$l start streamlink"
      nohup "$1" "$7" 1>/dev/null 2>/dev/null &

      #wait for steamlauncher to open
      until [[ $(pgrep steamlink) ]] ; do
          sleep 0.5
          log "$l wait steamlink"
      done

      #wait for steamlink to connect
      WAIT_NUMBER=0
      until [[ $(pidof streaming_client) ]] ; do
          sleep 0.5
          log "$l wait steamlink connected"
          WAIT_NUMBER=$((WAIT_NUMBER + 1))
          if [[ $WAIT_NUMBER -gt 120 ]] ; then
            log "$l waiting steamlink connect too long"
            startKodi
            break
          fi
      done

      #wait for steam to close
      STOP_NUMBER=0
      while [[ $STOP_NUMBER -lt 4 ]] ; do
          if [ $(pidof streaming_client) ] ; then
            STOP_NUMBER=0
          else
            STOP_NUMBER=$((STOP_NUMBER + 1))
          fi
          log "$l steam executable detected - looping"
          sleep 0.5
      done
      log "$l steam not detected any more - restarting kodi"

      #restarting/maximising kodi
      if [[ $6 != false ]] ; then
          "$6"
      fi
      startKodi

#####################################
        ;;
    *)
        log $l "I don't support this OS!"
        exit 1
        ;;
esac

#methods i have tried to detect between bpm minimised and hidden (eg when running a game)
#xwininfo -all -id $(wmctrl -lp | grep $(ps aux | grep -i 'ubuntu12_32/steam '| grep -v 'grep.*steam' | head -n1 | awk '{print $2}')| awk '{print $1}')
#xprop -id $(wmctrl -lp | grep $(ps aux | grep -i 'ubuntu12_32/steam '| grep -v 'grep.*steam' | head -n1 | awk '{print $2}')| awk '{print $1}')
#wmctrl -lpxG | grep $(ps aux | grep -i 'ubuntu12_32/steam '| grep -v 'grep.*steam' | head -n1 | awk '{print $2}')

resources/main.py
# steam launcher by teeedubb. http://forum.xbmc.org/showthread.php?tid=157499
# I used Rom Collection Browser as a guide when making this addon, plus borrowed ideas and code from it too. Big thanks to malte for RCB!
import os
import sys
import subprocess
import time
import shutil
import stat
import xbmc
import xbmcaddon
import xbmcgui

addon = xbmcaddon.Addon(id='script.steamlink.launcher')
addonPath = addon.getAddonInfo('path')
addonIcon = addon.getAddonInfo('icon')
addonVersion = addon.getAddonInfo('version')
dialog = xbmcgui.Dialog()
language = addon.getLocalizedString
scriptid = 'script.steamlink.launcher'

steamLinux = addon.getSetting("SteamLinux").decode("utf-8")
kodiLinux = addon.getSetting("KodiLinux").decode("utf-8")
steamWin = addon.getSetting("SteamWin").decode("utf-8")
kodiWin = addon.getSetting("KodiWin").decode("utf-8")
steamOsx = addon.getSetting("SteamOsx").decode("utf-8")
kodiOsx = addon.getSetting("KodiOsx").decode("utf-8")
delUserScriptSett = addon.getSetting("DelUserScript")
quitKodiSetting = addon.getSetting("QuitKodi")
busyDialogTime = int(addon.getSetting("BusyDialogTime"))
scriptUpdateCheck = addon.getSetting("ScriptUpdateCheck")
filePathCheck = addon.getSetting("FilePathCheck")
kodiPortable = addon.getSetting("KodiPortable")
preScriptEnabled = addon.getSetting("PreScriptEnabled")
preScript = addon.getSetting("PreScript").decode("utf-8")
postScriptEnabled = addon.getSetting("PostScriptEnabled")
postScript = addon.getSetting("PostScript").decode("utf-8")
osWin = xbmc.getCondVisibility('system.platform.windows')
osOsx = xbmc.getCondVisibility('system.platform.osx')
osLinux = xbmc.getCondVisibility('system.platform.linux')
osAndroid = xbmc.getCondVisibility('system.platform.android')
wmctrlCheck = addon.getSetting("WmctrlCheck")
suspendAudio = addon.getSetting("SuspendAudio")
customScriptFolder = addon.getSetting("CustomScriptFolder").decode("utf-8")
customScriptFolderEnabled = addon.getSetting("CustomScript")
minimiseKodi = addon.getSetting("MinimiseKodi")
steamParameters = addon.getSetting("SteamParameters")
forceKillKodi = addon.getSetting("ForceKillKodi")
desktopMode = addon.getSetting("DesktopMode")

def log(msg):
	msg = msg.encode(txt_encode)
	xbmc.log('%s: %s' % (scriptid, msg))

def getAddonInstallPath():
	path = addon.getAddonInfo('path').decode("utf-8")
	return path

def getAddonDataPath():
	path = xbmc.translatePath('special://profile/addon_data/%s' % scriptid).decode("utf-8")
	if not os.path.exists(path):
		log('addon userdata folder does not exist, creating: %s' % path)
		try:
			os.makedirs(path)
			log('created directory: %s' % path)
		except:
			log('ERROR: failed to create directory: %s' % path)
			dialog.notification(language(50212), language(50215), addonIcon, 5000)
	return path

def copyLauncherScriptsToUserdata():
	oldBasePath = os.path.join(getAddonInstallPath(), 'resources', 'scripts')
	if osWin:
		oldPath = os.path.join(oldBasePath, 'steam-launcher.ahk')
		newPath = os.path.join(scripts_path, 'steam-launcher.ahk')
		copyFile(oldPath, newPath)
		oldPath = os.path.join(oldBasePath, 'steam-launcher.exe')
		newPath = os.path.join(scripts_path, 'steam-launcher.exe')
		copyFile(oldPath, newPath)
	elif osLinux + osOsx:
		oldPath = os.path.join(oldBasePath, 'steam-launcher.sh')
		newPath = os.path.join(scripts_path, 'steam-launcher.sh')
		copyFile(oldPath, newPath)

def copyFile(oldPath, newPath):
	newDir = os.path.dirname(newPath)
	if not os.path.isdir(newDir):
		log('userdata scripts folder does not exist, creating: %s' % newDir)
		try:
			os.mkdir(newDir)
			log('sucsessfully created userdata scripts folder: %s' % newDir)
		except:
			log('ERROR: failed to create userdata scripts folder: %s' % newDir)
			dialog.notification(language(50212), language(50215), addonIcon, 5000)
			sys.exit()
	if not os.path.isfile(newPath):
		log('script file does not exist, copying to userdata: %s' % newPath)
		try:
			shutil.copy2(oldPath, newPath)
			log('sucsessfully copied userdata script: %s' % newPath)
		except:
			log('ERROR: failed to copy script file to userdata: %s' % newPath)
			dialog.notification(language(50212), language(50215), addonIcon, 5000)
			sys.exit()
	else:
		log('script file already exists, skipping copy to userdata: %s' % newPath)

def makeScriptExec():
	scriptPath = os.path.join(scripts_path, 'steam-launcher.sh')
	if os.path.isfile(scriptPath):
		if '\r\n' in open(scriptPath,'rb').read():
			log('Windows line endings found in %s, converting to unix line endings.' % scriptPath)
			with open(scriptPath, 'rb') as f:
				content = f.read()
				content = content.replace('\r\n', '\n')
			with open(scriptPath, 'wb') as f:
				f.write(content)
		if not stat.S_IXUSR & os.stat(scriptPath)[stat.ST_MODE]:
			log('steam-launcher.sh not executable: %s' % scriptPath)
			try:
				os.chmod(scriptPath, stat.S_IRWXU)
				log('steam-launcher.sh now executable: %s' % scriptPath)
			except:
				log('ERROR: unable to make steam-launcher.sh executable, exiting: %s' % scriptPath)
				dialog.notification(language(50212), language(50215), addonIcon, 5000)
				sys.exit()
			log('steam-launcher.sh executable: %s' % scriptPath)

def usrScriptDelete():
	if delUserScriptSett == 'true':
		log('deleting userdata scripts, option enabled: delUserScriptSett = %s' % delUserScriptSett)
		scriptFile = os.path.join(scripts_path, 'steam-launcher.ahk')
		delUserScript(scriptFile)
		scriptFile = os.path.join(scripts_path, 'steam-launcher.exe')
		delUserScript(scriptFile)
		scriptFile = os.path.join(scripts_path, 'steam-launcher.sh')
		delUserScript(scriptFile)
	elif delUserScriptSett == 'false':
		log('skipping deleting userdata scripts, option disabled: delUserScriptSett = %s' % delUserScriptSett)

def delUserScript(scriptFile):
	if os.path.isfile(scriptFile):
		try:
			os.remove(scriptFile)
			log('found and deleting: %s' % scriptFile)
		except:
			log('ERROR: deleting failed: %s' % scriptFile)
			dialog.notification(language(50212), language(50215), addonIcon, 5000)
		addon.setSetting(id="DelUserScript", value="false")

def fileChecker():
	if osLinux:
		if wmctrlCheck == 'true':
			if subprocess.call(["which", "wmctrl"]) != 0:
				log('ERROR: System program "wmctrl" not present, install it via you system package manager or if you are running the SteamOS compositor disable the addon option "Check for program wmctrl" (ONLY FOR CERTAIN USE CASES!!)')
				dialog.notification(language(50212), language(50215), addonIcon, 5000)
				sys.exit()
			else:
				log('wmctrl present, checking if a window manager is running...')
				if subprocess.call('DP=$(w -hs | awk \'{print $3}\') && DISPLAY=$DP wmctrl -l', shell=True) != 0:
					log('ERROR: A window manager is NOT running - unless you are using the SteamOS compositor Steam BPM needs a windows manager. If you are using the SteamOS compositor disable the addon option "Check for program wmctrl"')
					dialog.notification(language(50212), language(50215), addonIcon, 5000)
					sys.exit()
				else:
					log('A window manager is running...')
		if minimiseKodi == "true":
			if subprocess.call(["which", "xdotool"]) != 0:
				log('ERROR: Minimised Kodi enabled and system program "xdotool" not present, install it via you system package manager. Xdotool is required to minimise Kodi.')
				dialog.notification(language(50212), language(50215), addonIcon, 5000)
				sys.exit()
			else:
				log('xdotool present...')
	if filePathCheck == 'true':
		log('running program file check, option is enabled: filePathCheck = %s' % filePathCheck)
		if osWin:
			steamWin = addon.getSetting("SteamWin")
			kodiWin = addon.getSetting("KodiWin")
			steamExe = os.path.join(steamWin).decode("utf-8")
			xbmcExe = os.path.join(kodiWin).decode("utf-8")
			programFileCheck(steamExe, xbmcExe)
		elif osOsx:
			steamOsx = addon.getSetting("SteamOsx")
			kodiOsx = addon.getSetting("KodiOsx")
			steamExe = os.path.join(steamOsx).decode("utf-8")
			xbmcExe = os.path.join(kodiOsx).decode("utf-8")
			programFileCheck(steamExe, xbmcExe)
		elif osLinux:
			steamLinux = addon.getSetting("SteamLinux")
			kodiLinux = addon.getSetting("KodiLinux")
			steamExe = os.path.join(steamLinux).decode("utf-8")
			xbmcExe = os.path.join(kodiLinux).decode("utf-8")
			programFileCheck(steamExe, xbmcExe)
	else:
		log('skipping program file check, option disabled: filePathCheck = %s' % filePathCheck)

def fileCheckDialog(programExe):
	log('ERROR: dialog to go to addon settings because executable does not exist: %s' % programExe)
	if dialog.yesno(language(50212), programExe, language(50210), language(50211)):
		log('yes selected, opening addon settings')
		addon.openSettings()
		fileChecker()
		sys.exit()
	else:
		log('ERROR: no selected with invalid executable, exiting: %s' % programExe)
		sys.exit()

def programFileCheck(steamExe, xbmcExe):
	if osWin + osLinux:
		if os.path.isfile(os.path.join(steamExe)):
			log('Steam executable exists %s' % steamExe)
		else:
			fileCheckDialog(steamExe)
		if os.path.isfile(os.path.join(xbmcExe)):
			log('Xbmc executable exists %s' % xbmcExe)
		else:	
			fileCheckDialog(xbmcExe)
	if osOsx:
		if os.path.isdir(os.path.join(steamExe)):
			log('Steam folder exists %s' % steamExe)
		else:
			fileCheckDialog(steamExe)
		if os.path.isdir(os.path.join(xbmcExe)):
			log('Xbmc executable exists %s' % xbmcExe)
		else:	
			fileCheckDialog(xbmcExe)

def scriptVersionCheck():
	if scriptUpdateCheck == 'true':
		log('usr scripts are set to be checked for updates...')
		if delUserScriptSett == 'false':
			log('usr scripts are not set to be deleted, running version check')
			sysScriptDir = os.path.join(getAddonInstallPath(), 'resources', 'scripts')
			if osWin:
				sysScriptPath = os.path.join(sysScriptDir, 'steam-launcher.ahk')
				usrScriptPath = os.path.join(scripts_path, 'steam-launcher.ahk')
				if os.path.isfile(os.path.join(usrScriptPath)):
					compareFile(sysScriptPath, usrScriptPath)
				else:
					log('usr script does not exist, skipping version check')
			elif osLinux + osOsx:
				sysScriptPath = os.path.join(sysScriptDir, 'steam-launcher.sh')
				usrScriptPath = os.path.join(scripts_path, 'steam-launcher.sh')
				if os.path.isfile(os.path.join(usrScriptPath)):
					compareFile(sysScriptPath, usrScriptPath)
				else:
					log('usr script does not exist, skipping version check')
		else:
			log('usr scripts are set to be deleted, no version check needed')
	else:
		log('usr scripts are set to not be checked for updates, skipping version check')

def compareFile(sysScriptPath, usrScriptPath):
	global delUserScriptSett
	scriptSysVer = '000'
	scriptUsrVer = '000'
	if os.path.isfile(sysScriptPath):
		with open(sysScriptPath, 'r') as f:
			for line in f.readlines():
				if "steam.launcher.script.revision=" in line:
					scriptSysVer = line[32:35]
		log('sys "steam.launcher.script.revision=": %s' % scriptSysVer)
	if os.path.isfile(usrScriptPath):
		with open(usrScriptPath, 'r') as f:
			for line in f.readlines():
				if "steam.launcher.script.revision=" in line:
					scriptUsrVer = line[32:35]
		log('usr "steam.launcher.script.revision=": %s' % scriptUsrVer)
	if scriptSysVer > scriptUsrVer:
		log('system scripts have been updated: sys:%s > usr:%s' % (scriptSysVer, scriptUsrVer))
		if dialog.yesno(language(50113), language(50213), language(50214)):
			delUserScriptSett = 'true'
			log('yes selected, option delUserScriptSett enabled: %s' % delUserScriptSett)
		else:
			delUserScriptSett = 'false'
			log('no selected, script update check disabled: ScriptUpdateCheck = %s' % scriptUpdateCheck)
	else:
		log('userdata script are up to date')

def quitKodiDialog():
	global quitKodiSetting
	if quitKodiSetting == '2':
		log('quit setting: %s selected, asking user to pick' % quitKodiSetting)
		if dialog.yesno('Steam Launcher', language(50053)):
			quitKodiSetting = '0'
		else:
			quitKodiSetting = '1'
	if quitKodiSetting == '1' and minimiseKodi == "false":
		quitKodiSetting = '3'
	log('quit setting selected: %s' % quitKodiSetting)

def kodiBusyDialog():
	if busyDialogTime != 0:
		xbmc.executebuiltin("ActivateWindow(busydialognocancel)")
		log('busy dialog started')
		time.sleep(busyDialogTime)
		xbmc.executebuiltin("Dialog.Close(busydialognocancel)")
		log('busy dialog stopped after: %s seconds' % busyDialogTime)

def steamPrePost():
	global postScript
	global preScript
	if preScriptEnabled == 'false':
		preScript = 'false'
	elif preScriptEnabled == 'true':
		if not os.path.isfile(os.path.join(preScript)):
			log('pre-steam script does not exist, disabling!: "%s"' % preScript)
			preScript = 'false'
			dialog.notification(language(50212), language(50215), addonIcon, 5000)
	elif preScript == '':
		preScript = 'false'
	log('pre steam script: %s' % preScript)
	if postScriptEnabled == 'false':
		postScript = 'false'
	elif preScriptEnabled == 'true':
		if not os.path.isfile(os.path.join(postScript)):
			log('post-steam script does not exist, disabling!: "%s"' % postScript)
			postScript = 'false'
			dialog.notification(language(50212), language(50215), addonIcon, 5000)
	elif postScript == '':
		postScript = 'false'
	log('post steam script: %s' % postScript)

def launchSteam():
	if osAndroid:
		cmd = "com.valvesoftware.android.steam.community"
		log('attempting to launch: "%s"' % cmd)
		xbmc.executebuiltin('XBMC.StartAndroidActivity("%s")' % cmd)
		kodiBusyDialog()
		sys.exit()
	elif osWin:
		steamlauncher = os.path.join(scripts_path, 'steam-launcher.exe')
		cmd = '"%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s"' % (steamlauncher, steamWin, kodiWin, quitKodiSetting, kodiPortable, preScript, postScript, steamParameters, forceKillKodi, desktopMode)
	elif osOsx:
		steamlauncher = os.path.join(scripts_path, 'steam-launcher.sh')
		cmd = '"%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s"' % (steamlauncher, steamOsx, kodiOsx, quitKodiSetting, kodiPortable, preScript, postScript, steamParameters, forceKillKodi, desktopMode)
	elif osLinux:
		steamlauncher = os.path.join(scripts_path, 'steam-launcher.sh')
		cmd = '"%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" "%s" %s' % (steamlauncher, steamLinux, kodiLinux, quitKodiSetting, kodiPortable, preScript, postScript, steamParameters, forceKillKodi, desktopMode, "> " + steamlauncher.replace('.sh', '.log'))
	try:
		log('attempting to launch: %s' % cmd)
		print cmd.encode('utf-8')
		if suspendAudio == 'true':
			xbmc.audioSuspend()
			log('Audio suspended')
		if quitKodiSetting != '0' and suspendAudio == 'true':
			proc_h = subprocess.Popen(cmd.encode(txt_encode), shell=True, close_fds=False)
			kodiBusyDialog()
			log('Waiting for Steam to exit')
			while proc_h.returncode is None:
				xbmc.sleep(1000)
				proc_h.poll()
			log('Start resuming audio....')
			xbmc.audioResume()
			log('Audio resumed')
			del proc_h		
		else:
			subprocess.Popen(cmd.encode(txt_encode), shell=True, close_fds=True)
			kodiBusyDialog()
	except:
		log('ERROR: failed to launch: %s' % cmd)
		print cmd.encode(txt_encode)
		dialog.notification(language(50212), language(50215), addonIcon, 5000)

#HACK: sys.getfilesystemencoding() is not supported on all systems (e.g. Android)
txt_encode = 'utf-8'
try:
	txt_encode = sys.getfilesystemencoding()
except:
	pass
#osAndroid returns linux + android
if osAndroid: 
	osLinux = 0
	txt_encode = 'utf-8'
log('****Running Steam-Launcher v%s....' % addonVersion)
log('running on osAndroid, osOsx, osLinux, osWin: %s %s %s %s ' % (osAndroid, osOsx, osLinux, osWin))
log('System text encoding in use: %s' % txt_encode)
if customScriptFolderEnabled == 'true':
	scripts_path = customScriptFolder
else:
	scripts_path = os.path.join(getAddonDataPath(), 'scripts')
scriptVersionCheck()
usrScriptDelete()
copyLauncherScriptsToUserdata()
fileChecker()
makeScriptExec()
steamPrePost()
quitKodiDialog()
launchSteam()
